OpenRA.Mods.Common/Widgets/Logic/MapGeneratorLogic.cs

```diff
@@ -390,48 +401,61 @@ void OnClick()
 			}
 		}
 
+		System.Threading.CancellationTokenSource cancelationTokenSource;
+
 		void GenerateMap()
 		{
-			generating = true;
-			onGenerate(null, null);
+			var currentGeneration = ++generationCounter;
```

Comment:
Incrementing a volatile variable is not an atomic operation. Consider `Interlocked`.